### PR TITLE
Allow configuring cache folder

### PR DESCRIPTION
**Pull request by tomleb**
As briefly discussed on discord, this PR adds a config option to specify the cache directory. I added it as a CLI argument too (`--cachefolder`), as well as the usual config option (`CacheFolder`) and env var (`ND_CACHEFOLDER`).

This will allow linux distributions (eg: alpine, since I maintain the navidrome package on that one) to put the navidrome cache to `/var/cache/navidrome`.

I wasn't sure what default value to put between just `./cache` or `<datafolder>/cache`. Opted for `./cache` for now.

I haven't updated the website to add the new config option, I'll do once we agree on this MR / values, etc.

### CODE REVIEW

File: core/media_streamer_Internal_test.go [hunk 1]
```diff
@@ -20,7 +20,8 @@ var _ = Describe("MediaStreamer", func() {
 
 	BeforeEach(func() {
 		DeferCleanup(configtest.SetupConfig())
-		conf.Server.DataFolder, _ = os.MkdirTemp("", "file_caches")
+		conf.Server.DataFolder, _ = os.MkdirTemp("", "data")
+		conf.Server.CacheFolder, _ = os.MkdirTemp("", "file_caches")
```
Comment (deluan) [inline core/media_streamer_Internal_test.go:24]: Tests are passing without adding this line.

File: Makefile [hunk 2]
```diff
@@ -108,9 +108,9 @@ get-music: ##@Development Download some free music from Navidrome's demo instanc
 	mkdir -p music
 	( cd music; \
 	curl "https://demo.navidrome.org/rest/download?u=demo&p=demo&f=json&v=1.8.0&c=dev_download&id=ec2093ec4801402f1e17cc462195cdbb" > brock.zip; \
-	curl "https://demo.navidrome.org/rest/download?u=demo&p=demo&f=json&v=1.8.0&c=NavidromeUI&id=b376eeb4652d2498aa2b25ba0696725e" > back_on_earth.zip; \
-	curl "https://demo.navidrome.org/rest/download?u=demo&p=demo&f=json&v=1.8.0&c=NavidromeUI&id=e49c609b542fc51899ee8b53aa858cb4" > ugress.zip; \
-	curl "https://demo.navidrome.org/rest/download?u=demo&p=demo&f=json&v=1.8.0&c=NavidromeUI&id=350bcab3a4c1d93869e39ce496464f03" > voodoocuts.zip; \
+	curl "https://demo.navidrome.org/rest/download?u=demo&p=demo&f=json&v=1.8.0&c=dev_download&id=b376eeb4652d2498aa2b25ba0696725e" > back_on_earth.zip; \
```
Comment (deluan) [inline Makefile:111]: Nice catch!

File: core/media_streamer_test.go [hunk 3]
```diff
@@ -23,7 +23,8 @@ var _ = Describe("MediaStreamer", func() {
 
 	BeforeEach(func() {
 		DeferCleanup(configtest.SetupConfig())
-		conf.Server.DataFolder, _ = os.MkdirTemp("", "file_caches")
+		conf.Server.DataFolder, _ = os.MkdirTemp("", "data")
+		conf.Server.CacheFolder, _ = os.MkdirTemp("", "file_caches")
```
Comment (deluan) [inline core/media_streamer_test.go:27]: Same as above.

**Review by deluan — CHANGES_REQUESTED** [review]
Looks good, but we want the default to be `<datafolder>/cache` or else this will invalidate all caches from all users when this PR is merged. This could cause album covers to go missing for example.

**Comment by tomleb** [comment]
> Looks good, but we want the default to be `<datafolder>/cache` or else this will invalidate all caches from all users when this PR is merged. This could cause album covers to go missing for example.

Changed the behavior to default to `<datafolder>/cache`, let me know what you think. Though I think it would be better if the defaulting would be in conf/configuration.go's Load function. Seems like a simple thing to do but when I do that, tests fail. There seem to be a problem with the tests in `core`. eg:

```
• [FAILED] [0.001 seconds]
MediaStreamer NewStream [It] returns a NON seekable stream if transcode is required
/home/tomleb/dev/sources/navidrome/navidrome/core/media_streamer_test.go:56

  [FAILED] Expected
      <bool>: true
  to be false
  In [It] at: /home/tomleb/dev/sources/navidrome/navidrome/core/media_streamer_test.go:59 @ 05/19/23 21:13:25.983
```

My guess is there's an issue with `GetTranscodingCache()` which initializes the fs cache only once, even if `BeforeEach` sets a new `Server.DataFolder`. So first time it gets called, the fs cache points to some random dir, then next time it gets called, we create new temp dirs, but the fs cache picked up is the first one, aka the new temp dir is unused. Anyway, will check it out more tomorrow.

File: core/media_streamer_Internal_test.go (hunk 1 above)
Comment (tomleb) [reply core/media_streamer_Internal_test.go]: Removed those as suggested

File: core/media_streamer_test.go (hunk 3 above)
Comment (tomleb) [reply core/media_streamer_test.go]: Removed those as suggested

**Comment by deluan** [comment]
Thanks for fixing the tests! By the way, to force tests to run in order, use the [`Ordered` keyword in your tests](https://onsi.github.io/ginkgo/#ordered-containers).

Now that the tests are fixed, can you move the `CacheDir` defaulting to the `conf.Load()` function? If tests are still failing after moving it, push the changes and I can take a look.

**Comment by tomleb** [comment]
> Thanks for fixing the tests! By the way, to force tests to run in order, use the [`Ordered` keyword in your tests](https://onsi.github.io/ginkgo/#ordered-containers).

Oh thanks, didn't know about it.

> Now that the tests are fixed, can you move the `CacheFolder` defaulting to the `conf.Load()` function? If tests are still failing after moving it, push the changes and I can take a look.

I've moved the defaulting to `conf.Load()` as suggested. The tests are failing because that function isn't being called between runs so when `BeforeEach` changes the `DataFolder` value, it doesn't change the `CacheFolder` one. Do we want to instead set the CacheFolder value directly? Or find a way to call conf.Load() somehow?

I've left `MustPassRepeatedly` on purpose so that you can have a look with a non-working test ready.

**Comment by tomleb** [comment]
> The solution was to set CacheFolder directly, instead of DataFolder. Also there was an issue with the tests initialization that was causing the creating of a `cache` folder in the project's root 

Neat, what is the next step in terms of this PR? Can it be merged as is or is there things you want me to change?

**Comment by deluan** [comment]
I'll do a couple of tests with the new config, and then merge it. In the meantime, can you please document it in the documentation website?

**Pull request merged**